### PR TITLE
fix(logging): Fix sampling when log level is above Debug

### LINKED ISSUE #951: Bug: Sampler value always lower than sample rate
### Expected Behaviour

Think there is a bug when using log sampling. When I set POWERTOOLS_LOGGER_SAMPLE_RATE environment to 0.9 I would expect in 90 percent of the cases to have all logs elevated to debug level and thus logged even though minimum log level is set to Error using POWERTOOLS_LOG_LEVEL variable.

### Current Behaviour

Logs are never elevated to debug. It only happens when I set POWERTOOLS_LOGGER_SAMPLE_RATE to 1.

### Code snippet

```csharp
I am using the logger using the static log configuration in a console application and I set POWERTOOLS_LOG_LEVEL to Error and POWERTOOLS_LOGGER_SAMPLE_RATE to 0.9 using environment variables

[Logging(Service = "HelloWorldService", LoggerOutputCase = LoggerOutputCase.CamelCase, LogEvent = true)]
  static void Main(string[] args)
  {
  
    var logLevel = Environment.GetEnvironmentVariable("POWERTOOLS_LOG_LEVEL");
    Console.WriteLine($"POWERTOOLS_LOG_LEVEL: {logLevel}");
    var sampleRate = Environment.GetEnvironmentVariable("POWERTOOLS_LOGGER_SAMPLE_RATE");
    Console.WriteLine($"POWERTOOLS_SAMPLE_RATE: {sampleRate}");
  
    // This should NOT be logged (Info < Error)
    Logger.LogInformation("This is an info message — should not appear");

  }
```

### Possible Solution

_No response_

### Steps to Reproduce

Just create a console app using the same configuration and environment variables. Information is only loggen when sample rate is set to 1

### Powertools for AWS Lambda (.NET) version

2.0.0

### AWS Lambda function runtime

dotnet8

### PR DESCRIPTION
**Pull request by dcabib**
## Description

Fixes the GetSafeRandom() method in PowertoolsLoggerConfiguration.cs that was generating invalid values like 1.79E+308 instead of proper [0,1] range values for log sampling.

## Root Cause
The method was using BitConverter.ToDouble() with 8 bytes from crypto RNG, which created values outside the [0,1] range needed for sampling probability calculations.

## Solution
- Changed from BitConverter.ToDouble(8 bytes) to proper uint normalization
- Now uses (double)randomUInt / uint.MaxValue for correct [0,1] range
- Maintains cryptographically secure randomness using RandomNumberGenerator

## Changes Made
1. Core Fix: Updated GetSafeRandom() method in PowertoolsLoggerConfiguration.cs
2. Test Coverage: Added 4 comprehensive tests for sampling functionality:
   - Range validation [0,1]
   - Randomness diversity check
   - Integration test with real random generator
   - Edge case validation (zero sampling rate)

## Validation
All Core Components Passing (669/669 tests):
- Logging: 417/417 tests passed (including new sampling tests)
- Common: 67/67 tests passed
- Metrics: 86/86 tests passed  
- Tracing: 92/92 tests passed
- Metrics.AspNetCore: 7/7 tests passed

## Compatibility
- Works with both .NET 6 and .NET 8 targets
- No breaking changes
- Maintains existing API surface

Fixes #951

### CODE REVIEW

**Comment by hjgraca** [comment]
The current pull request, while technically correct, does not fully address the underlying issue. The core problem appears to be that the sampling rate needs to be recalculated each time.
Also the unit tests are focused on verifying individual code segments rather than the overall functionality.

Look at the TypeScript tests
https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/tests/unit/sampling.test.ts

And the TypeScript implementation recalculates the sampling rate as shown in this section of the code:
https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/packages/logger/src/Logger.ts#L585

To proceed, we should:

1. Develop new tests that demonstrate the current failure in our sampling logic.
2. Implement the necessary fixes based on the test results.
3. Ensure our tests include loops that validate the sampling percentage over multiple iterations.

File: examples/Logging/src/HelloWorld/SamplingTestFunction.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
+/*
```
Comment (hjgraca) [inline examples/Logging/src/HelloWorld/SamplingTestFunction.cs]: This file is not required, if we want to keep the file it cannot be a "Function", otherwise the Duplicate 'LambdaSerializer' attribute error will appear

**Comment by hjgraca** [comment]
@dcabib thanks for updating the pull request, currently the tests are failing.

**Comment by hjgraca** [comment]
Thanks @dcabib the tests are green now, I will run this pull request locally and update my findings

**Review by hjgraca — DISMISSED** [review]
LGTM

**Comment by hjgraca** [comment]
When the minimum log level was set above Debug (e.g., Error), the Microsoft.Extensions.Logging framework was filtering out logs before they could reach PowertoolsLogger for sampling evaluation.

## Problem
- Setting `POWERTOOLS_LOG_LEVEL=Error` and `POWERTOOLS_LOGGER_SAMPLE_RATE=0.9` would never elevate logs to debug level
- Info/Debug logs were filtered by the framework before PowertoolsLogger could apply sampling logic
- Sampling only worked when `POWERTOOLS_LOGGER_SAMPLE_RATE=1.0` (100%)

## Solution
- Modified `LoggerFactoryHelper.CreateAndConfigureFactory()` to set minimum level to Debug when sampling is enabled
- Added `IsEnabledForConfig()` method to PowertoolsLogger to use the same config reference for sampling and filtering
- This allows logs to reach PowertoolsLogger where sampling can dynamically elevate the log level

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
I think we need to update the documentation like this: https://docs.powertools.aws.dev/lambda/python/latest/core/logger/#sampling-debug-logs

Image

**Comment by hjgraca** [comment]
@leandrodamascena updated documentation

image

Also added a manual RefreshSampleRateCalculation method

**Review by hjgraca — APPROVED** [review]
GTM

**Review by leandrodamascena — APPROVED** [review]

**Comment by dcabib** [comment]
❤️
